### PR TITLE
[camera_avfoundation] fix stopVideoRecording waiting indefinitely and video lag at start

### PR DESCRIPTION
**Pull request by misos1**
Call to `startWriting` was moved to `startVideoRecordingWithCompletion` to fix situation when completion is not called when `_isRecording` is `YES` but `_videoWriter.status` is `AVAssetWriterStatusUnknown` in `stopVideoRecordingWithCompletion` which can happen if `stopVideoRecordingWithCompletion` is called right after `startVideoRecordingWithCompletion` and `didOutputSampleBuffer` had no chance to call `startWriting` in meantime and to avoid lag which happens when is called `startWriting` between first frame creation and its appending.

Fixes https://github.com/flutter/flutter/issues/132016
Fixes https://github.com/flutter/flutter/issues/151319

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

**Comment by misos1** [comment]
Btw integration test fails on my device because here `file.path` ends with `CAP_D5507C07-56E6-48E5-96F8-7D22FF8B5AB6.jpg` but `fileFormat.name` is `jpeg` so file path does not end with file format name.

```
flutter: 00:55 +8: Capture specific image output formats
camera_avfoundation/example/integration_test/camera_test.dart:339
> expect(file.path.endsWith(fileFormat.name), true);
```

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:72]: nit: isFirstSample

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:848]: why is this check removed?

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:835]: do you intend to remove `startWriting` from `didOutputSampleBuffer`?

**Comment by hellohuanlin** [comment]
> Btw integration tests fail on my side with flutter packages main branch because here `file.path` for example ends with `CAP_D5507C07-56E6-48E5-96F8-7D22FF8B5AB6.jpg` but `fileFormat.name` is `jpeg` so file path does not end with file format name.

Seems reasonable to allow both jpg and jpeg to pass

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (misos1) [reply packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:835]: yes

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (misos1) [reply packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:848]: This is where it was not sending completion (in false case) so a dart future never resolved. But now `startWriting` is called at the same time as `_isRecording` is set to `YES` so `finishWritingWithCompletionHandler` in `stopVideoRecordingWithCompletion` can be called only after was called `startWriting` which sets `_videoWriter.status` to `AVAssetWriterStatusWriting` so that check redundant (and again it would just indicate false possibility of not calling completion).

Comment (misos1) [reply packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:848]: Actually maybe it should check whether status is `AVAssetWriterStatusWriting` and complete with error otherwise because it can be for example `AVAssetWriterStatusFailed` here (I suppose `finishWritingWithCompletionHandler ` would just crash in that case). But this situation is nothing new introduced with this PR as before there was just check `status != unknown` which would pass `failed` to `finishWritingWithCompletionHandler`.

I also noticed check for `AVAssetWriterStatusFailed` in `didOutputSampleBuffer` where is called `[self reportErrorMessage]` but when I tried to force call `reportErrorMessage` I did not see this error report anywhere.

Comment (misos1) [reply packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:852]: Regarding previous comment, seems `finishWritingWithCompletionHandler` is ok with `AVAssetWriterStatusFailed` it will just pass it to handler so there is no need to check whether is status not `failed` before calling it, it only has problem with `AVAssetWriterStatusCancelled` but there is no call to `cancelWriting`, `AVAssetWriterStatusCompleted` and `AVAssetWriterStatusUnknown` but these two also cannot happen.

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m:72]: Should this be `isFirstVideoSample`?

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m:834]: nit: "do not move startWriting call to didOutputSampleBuffer... "

Because there should only be 1 place to call `startWriting`.

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m:836]: can you also explain that this happens when stop video is called immediately after start, before the first video sample is arrived (if i understand correctly)

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m:852]: can you add a comment explaining that it's not necessary to check `AVAssetWriterStatusUnknown` anymore since we call startWriting immediately, even before video sample callback (please rephrase this).

**Review by hellohuanlin — COMMENTED** [review]
I think it's reasonable change. Just some nit to make it easier for future readers.

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m
Comment (misos1) [reply packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m:852]: ok, but such a comment loses context here because in sources it is not visible that there was such a check, only in this diff

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m:867]: nit: "so there's no need to check `_videoWriter.status` before finish writing"

**Review by hellohuanlin — APPROVED** [review]

**Comment by misos1** [comment]
Btw does this also require a second review?

**Comment by hellohuanlin** [comment]
> Btw does this also require a second review?

Only first time contributors require a second review, so you are good to go. But feel free to tag more people as you see fit.

**Comment by misos1** [comment]
Oh ok (there are comments in my 2nd and 3rd PR that second reviews are required, I did my first some year ago).

**Comment by hellohuanlin** [comment]
> Only first time contributors require a second review, so you are good to go. But feel free to tag more people as you see fit.

Sorry I was terribly wrong - we need a second reviewer to land this.

**Comment by jmagman** [comment]
For the curious, from https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#getting-a-code-review

> Every PR must be code-reviewed before check-in, including things like rolling a dependency. Getting a review means that a regular Flutter contributor (someone with commit access; see [contributor access](https://github.com/flutter/flutter/blob/master/docs/contributing/Contributor-access.md) for details) has "approved" the PR in the GitHub UI. We call this "getting an LGTM" ("looks good to me").
> 
> If you are not yourself someone with commit access, then a second person with commit access must also review and approve your PR. This ensures that two people with commit access (trusted contributors) agree on every commit.

**Review by bparrishMines — APPROVED** [review]
LGTM

**Comment by hellohuanlin** [comment]
There's also a process to become a member: https://github.com/flutter/flutter/blob/master/docs/contributing/Contributor-access.md

If you are interested, I'm more than happy to sponsor it.

**Comment by misos1** [comment]
@hellohuanlin Ok so let's try it.

**Pull request merged**
